### ISSUE TITLE
`torch.cumsum()` on bfloat16 tensor with dtype=torch.int8 produce inconsistent results between CPU and GPU

### ISSUE THREAD

**Issue by SilentTester73** [pytorch/pytorch#153359]
### 🐛 Describe the bug

When using torch.cumsum(tensor, dim=-1, dtype=torch.int8) on an input tensor of dtype=torch.bfloat16, the results differ between CPU and GPU. This may suggest an issue with the internal type conversion from bfloat16 to int8 within the cumsum operation when the dtype parameter is specified.

```python
import torch

# Input bfloat16 tensor
tensor_bf16 = torch.tensor([
    [ 1.0986e-25, -1.9114e+30, -9.5291e-20, -4.3163e-39, -1.2337e+23,
      2.0698e+30,  1.1934e+30],
    [-4.3910e-18, -4.3910e-18, -4.3910e-18, -4.3910e-18,  1.1934e+30,
      1.1934e+30, -4.4256e+28],
    [-4.9304e-29, -3.2969e-11, -8.0691e+05, -8.8500e+08, -2.0938e+00,
     -2.4629e+16,  5.6820e+26]], dtype=torch.bfloat16)

# Cumsum with dtype=torch.int8 on CPU
cpu_res = torch.cumsum(tensor_bf16, -1, dtype=torch.int8)

# Cumsum with dtype=torch.int8 on GPU
# Ensure CUDA is available before running this part
if torch.cuda.is_available():
    gpu_res = torch.cumsum(tensor_bf16.to("cuda"), -1, dtype=torch.int8)
    gpu_res_cpu = gpu_res.to("cpu")
    difference = cpu_res - gpu_res_cpu
else:
    gpu_res = "CUDA not available"
    gpu_res_cpu = "CUDA not available"
    difference = "CUDA not available"

print("Input bfloat16 tensor:")
print(tensor_bf16)

print("\nCPU cumsum result (dtype=torch.int8):")
print(cpu_res)

print("\nGPU cumsum result (dtype=torch.int8) (moved to CPU for printing):")
print(gpu_res_cpu)

print("\nDifference (CPU - GPU):")
print(difference)

# For calculating expected behavior:
# 1. Explicitly cast bfloat16 tensor to int8
tensor_int8_explicit = tensor_bf16.to(torch.int8)
# 2. Perform cumsum on the explicitly casted int8 tensor
expected_res = torch.cumsum(tensor_int8_explicit, -1, dtype=torch.int8) # dtype ensures int8 accumulation

print("\n--- Expected Behavior Analysis ---")
print("\nInput tensor explicitly cast to int8:")
print(tensor_int8_explicit)
print("\nExpected cumsum result (after explicit cast to int8):")
print(expected_res)
```

```output
Input bfloat16 tensor:
tensor([[ 1.0986e-25, -1.9114e+30, -9.5291e-20, -4.3163e-39, -1.2337e+23,
          2.0698e+30,  1.1934e+30],
        [-4.3910e-18, -4.3910e-18, -4.3910e-18, -4.3910e-18,  1.1934e+30,
          1.1934e+30, -4.4256e+28],
        [-4.9304e-29, -3.2969e-11, -8.0691e+05, -8.8500e+08, -2.0938e+00,
         -2.4629e+16,  5.6820e+26]], dtype=torch.bfloat16)

CPU cumsum result (dtype=torch.int8):
tensor([[ 0,  0,  0,  0,  0,  0,  0],
        [ 0,  0,  0,  0,  0,  0,  0],
        [ 0,  0,  0,  0, -2, -2, -2]], dtype=torch.int8)

GPU cumsum result (dtype=torch.int8) (moved to CPU for printing):
tensor([[ 0,  0,  0,  0,  0, -1, -2],
        [ 0,  0,  0,  0, -1, -2, -2],
        [ 0,  0,  0,  0, -2, -2, -3]], dtype=torch.int8)

Difference (CPU - GPU):
tensor([[0, 0, 0, 0, 0, 1, 2],
        [0, 0, 0, 0, 1, 2, 2],
        [0, 0, 0, 0, 0, 0, 1]], dtype=torch.int8)

--- Expected Behavior Analysis ---

Input tensor explicitly cast to int8:
tensor([[ 0,  0,  0,  0,  0,  0,  0],
        [ 0,  0,  0,  0,  0,  0,  0],
        [ 0,  0,  0,  0, -2,  0,  0]], dtype=torch.int8)

Expected cumsum result (after explicit cast to int8):
tensor([[ 0,  0,  0,  0,  0,  0,  0],
        [ 0,  0,  0,  0,  0,  0,  0],
        [ 0,  0,  0,  0, -2, -2, -2]], dtype=torch.int8)

```

Colab: https://colab.research.google.com/drive/1_yR5BMhV-n6wpDKXROK9LcEuLIGgJYLA?usp=sharing

### Versions

Collecting environment information...
PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: 18.1.8 (++20240731025043+3b5b5c1ec4a3-1~exp1~20240731145144.92)
CMake version: version 4.0.0
Libc version: glibc-2.39

Python version: 3.12.3 (main, Feb  4 2025, 14:48:35) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 570.133.20
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i9-13900F
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            1
Stepping:                             1
CPU(s) scaling MHz:                   27%
CPU max MHz:                          5600.0000
CPU min MHz:                          800.0000
BogoMIPS:                             3993.60
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            896 KiB (24 instances)
L1i cache:                            1.3 MiB (24 instances)
L2 cache:                             32 MiB (12 instances)
L3 cache:                             36 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.5
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] optree==0.15.0
[pip3] torch==2.7.0
[pip3] triton==3.3.0
[conda] Could not collect

**Comment by ngimel** [comment]
Identical results are not expected between cpu and gpu.
